### PR TITLE
HDDS-2114: Rename does not preserve non-explicitly created interim directories

### PR DESCRIPTION
**Pull request by lokeshj1703**

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 36 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| -1 | mvninstall | 31 | hadoop-ozone in trunk failed. |
| -1 | compile | 19 | hadoop-ozone in trunk failed. |
| +1 | checkstyle | 64 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 896 | branch has no errors when building and testing our client artifacts. |
| -1 | javadoc | 13 | hadoop-hdds in trunk failed. |
| -1 | javadoc | 13 | hadoop-ozone in trunk failed. |
| 0 | spotbugs | 159 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| -1 | findbugs | 22 | hadoop-ozone in trunk failed. |
||| _ Patch Compile Tests _ |
| -1 | mvninstall | 30 | hadoop-ozone in the patch failed. |
| -1 | compile | 20 | hadoop-ozone in the patch failed. |
| -1 | javac | 20 | hadoop-ozone in the patch failed. |
| -0 | checkstyle | 26 | hadoop-ozone: The patch generated 2 new + 121 unchanged - 0 fixed = 123 total (was 121) |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 716 | patch has no errors when building and testing our client artifacts. |
| -1 | javadoc | 13 | hadoop-hdds in the patch failed. |
| -1 | javadoc | 13 | hadoop-ozone in the patch failed. |
| -1 | findbugs | 21 | hadoop-ozone in the patch failed. |
||| _ Other Tests _ |
| -1 | unit | 137 | hadoop-hdds in the patch failed. |
| -1 | unit | 24 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 2860 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.container.keyvalue.TestKeyValueContainer |
|   | hadoop.ozone.container.ozoneimpl.TestOzoneContainer |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1440 |
| JIRA Issue | HDDS-2114 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux c3b727f210a2 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 39e82ac |
| Default Java | 1.8.0_222 |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/branch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/branch-compile-hadoop-ozone.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/branch-javadoc-hadoop-hdds.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/branch-javadoc-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/branch-findbugs-hadoop-ozone.txt |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/patch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/patch-compile-hadoop-ozone.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/patch-compile-hadoop-ozone.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/diff-checkstyle-hadoop-ozone.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/patch-javadoc-hadoop-hdds.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/patch-javadoc-hadoop-ozone.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/patch-findbugs-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/testReport/ |
| Max. process+thread count | 306 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/ozonefs U: hadoop-ozone/ozonefs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1440/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneFileSystem.java [hunk 1]
```diff
@@ -374,7 +374,11 @@ public boolean rename(Path src, Path dst) throws IOException {
       }
     }
     RenameIterator iterator = new RenameIterator(src, dst);
-    return iterator.iterate();
+    boolean result = iterator.iterate();
+    if (result) {
+      createFakeParentDirectory(src);
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneFileSystem.java]: should we createFakeParentDirectory for dst as it is the rename result?

**Comment by anuengineer** [comment]
I am going to +1 this. Since we want to make sure Hive works.

I just want to understand this more clearly. The issue is really that if we were a real file system, then there is nothing called an implicit path. Since we are an object store, there is a notion of a implicitly created file system. I am guess that S3AFS has the same problem, and either Hive has a workaround for this, or S3A is doing something really clever. Do we know how Hive works on S3?

**Comment by lokeshj1703** [comment]
@anuengineer Yes, its the same problem as you described. After rename if the sources's parent has no remaining children, then S3A does a mkdir for the parent. The PR does exactly the same for ozonefs.

**Pull request closed without merging**